### DISCUSSION TITLE
XSPICE Digital Library and Example Projects

### DISCUSSION THREAD

**Discussion by tomhajjar** [ra3xdh/qucs_s#1155] [Show and tell]
Updated 1/21/25

Major update. The new capabilities of the "SPICE netlist" and "SPICE Library" devices now allows the "Create Library..." Utility to create Libraries with Parameter Passing and embedded Spice models.

I have created six XSPICE Digital/Analog mixed-signal projects and the subcircuits for new Digital/Analog XSPICE Libraries. The library(s) have the basic Digital gates, Flip Flops, ADC/DAC Bridges, pull-up/dn and multiple Digital and Analog Clock sources and will be expanded over time to add more devices. Also done were some "X-Analog" subcircuits like OpAmps used in mixed-mode simulations. The XSPICE models are "bare-bones" to speed simulation times.

The projects will be a good primer for using XSPICE to simulated mixed digital-analog circuits. XSPICE ADC/DAC Bridges allow XSPICE Digital devices to be embedded in almost any simulation. For example the ngspice XSPICE VCO d_osc is very useful in analog simulations. 

Included are the projects to create the Library, XSPICE PLL and Delta Sigma Converter examples, XSPICE PLL Clock Recovery project using a Hogge Phase Detector, XSPICE mixed-signal example C3 and an XSPICE PRBS15 source done by @clyde73 who did an XSPICE 555 timer for us.

XSPICE digital components default to vcc=1 unless a .Param vcc=xx is used. All nodes are "digital" unless a DAC bridge is used or a connection is made to a "analog" device or resistor which initiates an "Auto-Bridge". I made 1/4/8 ADC-DAC Bridges and resistor packs to speed making schematics and add "digital" and "analog" interfaces between component types.

Plotting of voltages at "Digital" nodes requires using the "Plot Vs." option. Both "Digital "and "Analog" data can be plotted on the same Graph.

Simulations are faster if minimal Auto-Bridges are initiated but for typical circuits speed isn't an issue. I haven't tried making subcircuits to implement the "LTspice" method of plotting digital timing diagrams. Just using equations as shown in the examples.

![2024-12-17_154349](https://github.com/user-attachments/assets/e58bea3c-fd25-40b1-b17e-94288afbab7c)
![2024-12-11_190015](https://github.com/user-attachments/assets/7171adfe-7323-48e3-8c4f-8761d95e39ee)
![2024-12-11_194048](https://github.com/user-attachments/assets/d0708092-cef7-45c3-9425-d855f8d5573b)
![2024-12-22_124429](https://github.com/user-attachments/assets/a80e6588-7ea7-4667-86e1-3a988857b7cf)
![2024-12-22_154401](https://github.com/user-attachments/assets/7925b0f3-8b1c-4adf-b31f-1005ab5b92b0)
![2024-12-22_131458](https://github.com/user-attachments/assets/1287172c-aaf1-48d7-813c-97b9651dc3d2)
![2024-12-25_192532](https://github.com/user-attachments/assets/5d7185e1-be38-4570-89da-c3758e759c4d)
![2025-01-07_192636](https://github.com/user-attachments/assets/49c002ac-0efe-4785-9f20-7cb26a43153d)

[XSPICE_Mixed-Signal_Projects.zip](https://github.com/user-attachments/files/18493849/XSPICE_Mixed-Signal_Projects.zip)

**Comment by ra3xdh** [comment]
I have moved this to the discussions and marked with *examples* label. This autoamtically adds this thread to the examples list in this pinned discussion: https://github.com/ra3xdh/qucs_s/discussions/788

**Comment by ra3xdh** [comment]
Thanks for the contribution! Now it is added in the examples list.

**Comment by tomhajjar** [comment]
@ra3xdh 

The XSPICE PLL project was missing from above so I attached the latest update.
